src/com/ms/silverking/net/async/time/RandomBackoff.java

```diff
@@ -1,130 +1,121 @@
+      throw new RuntimeException("Bad extraBackoffValue: " + extraBackoffValue);
     }
-    
-    public int backoff(Object waitObj) {
-        return backoff(waitObj, true);
+    this.extraBackoffValue = extraBackoffValue;
+    this.exponential = exponential;
+    this.hardDeadline = hardDeadline;
+    random = new Random(seed);
+  }
+
+  public RandomBackoff(int maxBackoffNum, int initialBackoffValue, int extraBackoffValue, long hardDeadline) {
+    this(maxBackoffNum, initialBackoffValue, extraBackoffValue, hardDeadline, true,
+        SystemTimeUtil.skSystemTimeSource.absTimeMillis());
+  }
+
+  public RandomBackoff(int maxBackoffNum, int initialBackoffValue, long hardDeadline) {
+    this(maxBackoffNum, initialBackoffValue, 0, hardDeadline);
+  }
+
+  public void reset() {
+    curBackoffNum = 0;
+    consecutiveBackoffs = 0;
+  }
+
+  protected int getCurBackoffLimit(int backoffNum) {
+    if (exponential) {
+      return initialBackoffValue << backoffNum;
+    } else {
+      return initialBackoffValue * (backoffNum + 1);
     }
-    
-    public int backoff(Object waitObj, boolean waitHere) {
-        int    backoffTime;
-        
-        backoffTime = getRandomBackoff();
-        curBackoffNum = Math.min(curBackoffNum + 1, maxBackoffNum);
-        //System.out.println(curBackoffNum +"\t"+ maxBackoffNum +"\t"+ backoffTime);
-        consecutiveBackoffs++;
-        if (waitHere) {
-            if (hardDeadline > 0) {
-                backoffTime = (int)Math.min(backoffTime, hardDeadline - SystemTimeSource.instance.absTimeMillis());
-            }
-            if (waitObj == null) {
-                ThreadUtil.sleep(backoffTime);
-            } else {
-                synchronized (waitObj) {
-                    try {
-                        waitObj.wait(backoffTime);
-                    } catch (InterruptedException ie) {
-                    }
-                }
-            }
+  }
+
+  public int getRandomBackoff(int backoffNum) {
+    return random.nextInt(getCurBackoffLimit(backoffNum)) + extraBackoffValue;
+  }
+
+  protected int getRandomBackoff() {
+    return random.nextInt(getCurBackoffLimit(curBackoffNum)) + extraBackoffValue;
+  }
+
+  public int backoff(Object waitObj) {
+    return backoff(waitObj, true);
+  }
+
+  public int backoff(Object waitObj, boolean waitHere) {
+    int backoffTime;
+
+    backoffTime = getRandomBackoff();
+    curBackoffNum = Math.min(curBackoffNum + 1, maxBackoffNum);
+    //System.out.println(curBackoffNum +"\t"+ maxBackoffNum +"\t"+ backoffTime);
+    consecutiveBackoffs++;
+    if (waitHere) {
+      if (hardDeadline > 0) {
+        backoffTime = (int) Math.min(backoffTime, hardDeadline - SystemTimeUtil.skSystemTimeSource.absTimeMillis());
+      }
+      if (waitObj == null) {
+        ThreadUtil.sleep(backoffTime);
+      } else {
+        synchronized (waitObj) {
+          try {
+            waitObj.wait(backoffTime);
+          } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)